WordPress/openverse-api
# Add command to resend verification emails

## Fixes


Fixes #689 by @AetherUnbound  

## Description


Adds a new command to run for resending verification emails. Note: I haven't written unit tests for this but I need to stop working on this for the day and come back to it tomorrow. Just wanted to get a draft up early.

## Testing Instructions

I tried to write comprehensive unit tests. If you notice a missing branch please let me know so I can add additional tests.

You can test this on your local API (though I'm hoping the unit tests are sufficient) by setting the email backend to SMTP in your API settings without filling in the SMTP server information and then making some registration requests with multiple application names. Use a mix of the same email and different emails. This will cause the verifications to be completed, but because we haven't filled in the SMTP server information, it will fail to send the email (which is what was happening in production).

Then switch your settings back to the Console email backend and run the management command. You should see some helpful logs throughout the way, but also you'll see the emails that _would_ be sent pop into the console. It will look like this:

```
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: Verify your API credentials
From: 
To: user@example.com
Date: Thu, 12 May 2022 09:20:50 -0000
Message-ID: <165234725049.145.11337592980239670284@fa90ab837821>


The Openverse API OAuth2 email verification process has recently been fixed.
We have detected that you attempted to register an application using this email.

To verify your Openverse API credentials, click on the following link:

https://api.openverse.engineering//v1/auth_tokens/verify/<token redacted>

If you believe you received this message in error, please disregard it.

-------------------------------------------------------------------------------
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: Verify your API credentials
From: 
To: user@example.net
Date: Thu, 12 May 2022 09:20:50 -0000
Message-ID: <165234725049.145.2282716428617802226@fa90ab837821>


The Openverse API OAuth2 email verification process has recently been fixed.
We have detected that you attempted to register an application using this email.

To verify your Openverse API credentials, click on the following link:

https://api.openverse.engineering//v1/auth_tokens/verify/<token redacted>

If you believe you received this message in error, please disregard it.

-------------------------------------------------------------------------------
The following applications had email verifications sent.                                         

Application name: My amazing project
Cleaned related application count: 2
Cleaned related verification count: 0
Cleaned related registration count: 2


Application name: My amazing project5
Cleaned related application count: 0
Cleaned related verification count: 0
Cleaned related registration count: 0
```

To run the command:

For a dry run:
```
just dj resendoauthverification
```

For an actual run:
```
just dj resendoauthverification --no-dry_run
```

Note that the non-dry run requires you to input exactly `YES` as the response. Any other input including `yes` will not be accepted.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
